source/include/TypeAndDataManager.h

```diff
@@ -366,20 +366,39 @@ template <typename TSource, typename TDest>
 TDest ConvertFloatToInt(TSource src)
 {
     TDest dest;
+    TSource maxDestValue = static_cast<TSource>(std::numeric_limits<TDest>::max());
     if (std::isnan(src)) {
         dest = std::numeric_limits<TDest>::max();
     } else if (std::isinf(src)) {
         dest = src < 0 ? std::numeric_limits<TDest>::min() : std::numeric_limits<TDest>::max();
     } else if (src < std::numeric_limits<TDest>::min()) {
         dest = std::numeric_limits<TDest>::min();
-    } else if (src > std::numeric_limits<TDest>::max()) {
+    } else if (src > maxDestValue) {
         dest = std::numeric_limits<TDest>::max();
     } else {
         dest = static_cast<TDest>(RoundToEven(src));
     }
     return dest;
 }
 
+template <typename TDest>
+TDest ConvertDoubleToInt(Double src)
+{
+    TDest dest;
```

Comment:
Curious. How did you come up with this logic? Was it documented in emscripten? I don't quite understand use of fmod. Could you explain?

Reply:
As mentioned emscrripten site BINARYEN_TRAP_MODE=js  was the same as how JS handling the overflow and underflow. From the [emca ](https://262.ecma-international.org/6.0/#sec-touint8)site, we can see how ToInt32 ToInt8 etc works. I tried to get similar behavior in C++ with this function.
 By default, now it traps overflow and set it to the highest possible value and for underflow, it takes the lowest supported value.  With the latest emsdk update [this ](https://github.com/ni/VireoSDK/blob/c8eb11d197b09f7ed2100bdb01cd48cd8f2fa003/test-it/karma/publicapi/Double.Test.js#L140)vireo test was failing. We were getting mismatched values from JS Typed Array and WASM code. With this change, it is fixed now.

Reply:
The goal of the Double.Test.js was to make sure that writing values to memory using the writeDouble api and by getting access to the array buffer in memory behave the same. The writeDouble api does the float to int conversion in C++ while accessing the array buffer relies on the JS float to int conversion using the spec link devika gave.

Looking at how Vireo does float to int on the diagram shows that [Vireo saturates using ConvertFloatToInt](https://github.com/ni/VireoSDK/blob/c8eb11d197b09f7ed2100bdb01cd48cd8f2fa003/source/include/TypeAndDataManager.h#L366) instead which seems to align with NXG desktop with a quick test. @sanmut is this intentionally defined this way?

If that's the case then the test might be backwards. The WriteDoubleToMemory function should probably be calling ConvertFloatToInt instead of having it's own implementation of float to int (either the casting from before or the new approach mimicking the JS behavior). And the JS side should have a function available to help with float to int conversion when assigning to type array and should not assign values directly to the typed array.

I plan to discuss it with @sanmut in the morning to see if that makes sense.

Reply:
I agree with @rajsite. I confirmed the behavior of saturation of floating point to integral conversion with LabVIEW itself. It is better to use the existing "ConvertFloatToInt" as Milan described above.  We discussed this over a meeting with @Devika-Tantry and @rajsite to make the suggested changes in the code and modify the test.

Reply:
Done. Used `ConvertFloatToInt` for Double to Int conversion and removed all references to the Typed array in the test.